File: app/src/main/java/org/wikipedia/analytics/LoginFunnel.java
package org.wikipedia.analytics;

import org.wikipedia.WikipediaApp;

/**
 * Schema: https://meta.wikimedia.org/wiki/Schema:MobileWikiAppLogin
 */
public class LoginFunnel extends Funnel {
    private static final String SCHEMA_NAME = "MobileWikiAppLogin";
    private static final int REVISION = 18144592;

    public static final String SOURCE_NAV = "navigation";
    public static final String SOURCE_EDIT = "edit";
    public static final String SOURCE_BLOCKED = "blocked";
    public static final String SOURCE_SYSTEM = "system";
    public static final String SOURCE_ONBOARDING = "onboarding";
    public static final String SOURCE_SETTINGS = "settings";
    public static final String SOURCE_READING_MANUAL_SYNC = "reading_lists_manual_sync";

    public LoginFunnel(WikipediaApp app) {
        super(app, SCHEMA_NAME, REVISION);
    }

    public void logStart(String source) {
        logStart(source, null);
    }

    public void logStart(String source, String editSessionToken) {
        log(
                "action", "start",
                "source", source,
                "edit_session_token", editSessionToken
        );
    }

    public void logCreateAccountAttempt() {
        log(
                "action", "createAccountAttempt"
        );
    }

    public void logCreateAccountFailure() {
        log(
                "action", "createAccountFailure"
        );
    }

    public void logCreateAccountSuccess() {
        log(
                "action", "createAccountSuccess"
        );
    }

    public void logError(String code) {
        log(
                "action", "error",
                "error_text", code
        );
    }

    public void logSuccess() {
        log(
                "action", "success"
        );
    }
}

File: app/src/main/java/org/wikipedia/analytics/LoginFunnel.kt
package org.wikipedia.analytics

import org.wikipedia.WikipediaApp

class LoginFunnel(app: WikipediaApp) : Funnel(app, SCHEMA_NAME, REVISION) {
    fun logStart(source: String = "", editSessionToken: String = "") {
        log("action" to "start", "source" to source, "edit_session_token" to editSessionToken)
    }

    fun logCreateAccountAttempt() {
        log("action" to "createAccountAttempt")
    }

    fun logCreateAccountFailure() {
        log("action" to "createAccountFailure")
    }

    fun logCreateAccountSuccess() {
        log("action" to "createAccountSuccess")
    }

    fun logError(code: String?) {
        log("action" to "error", "error_text" to code)
    }

    fun logSuccess() {
        log("action" to "success")
    }

    companion object {
        private const val SCHEMA_NAME = "MobileWikiAppLogin"
        private const val REVISION = 20710032
        const val SOURCE_NAV = "navigation"
        const val SOURCE_EDIT = "edit"
        const val SOURCE_BLOCKED = "blocked"
        const val SOURCE_SYSTEM = "system"
        const val SOURCE_ONBOARDING = "onboarding"
        const val SOURCE_SETTINGS = "settings"
        const val SOURCE_READING_MANUAL_SYNC = "reading_lists_manual_sync"
        const val SOURCE_LOGOUT_BACKGROUND = "logout_background"
        const val SOURCE_SUGGESTED_EDITS = "suggestededits"
    }
}
